### PR TITLE
Implements #145 - Auth flow 

### PR DESCRIPTION
**Pull request by manangouhari**
**What does this PR do?**
Implements the whole authentication flow.
Also handles these edge cases - 
1. Browser not supported.
2. Thanos not installed.
3. Account not verified.
4. Nickname or email already registered with another account.

Requesting you to let me know if I've missed something over here.

**How should this be manually tested?**
It can't be tested through the preview URL because the backend isn't hosted yet.
On your local machine -
1. Get the backend up and running. 
2. On the frontend, go to `/auth` and test the flow. 
3. To test a few edge cases - try going to `/auth` on Safari(Browser not supported) and Incognito tab on Brave(Thanos not available).

**Any background context you want to provide?**
A few things that need to be implemented to make this robust(will be done later in a different PR) - 
1. Store auth state in `localStorage` to persist login between browser refreshes.
2. Logic to check if the user is logged in to run on each page for auto-login.

### CODE REVIEW

**Comment by bhaskarSingh** [comment]
` if (windowLocalStorageAvailable()) {
8:18:22 PM:   782 |       localStorage = window.localStorage;
8:18:22 PM: > 783 |     } else if (window.globalStorage) {
8:18:22 PM:       |      ^
8:18:22 PM:   784 |       // Firefox 2-3 use globalStorage
8:18:22 PM:   785 |       // See https://developer.mozilla.org/en/dom/storage#globalStorage
8:18:22 PM:   786 |       try {
8:18:22 PM: 
8:18:22 PM:   WebpackError: ReferenceError: window is not defined
`

Somewhere window object being is used before the first load. 

solution would be to add condition eg:

`if(window !== undefined) { //proceeed }`

@manangouhari When the build is failing, the first thing you should do is check the build preview(https://app.netlify.com/sites/blissful-mcnulty-82f253/deploys/6012cd6488eb220007562ef5).

**Comment by manangouhari** [comment]
> @manangouhari When the build is failing, the first thing you should do is check the build preview(https://app.netlify.com/sites/blissful-mcnulty-82f253/deploys/6012cd6488eb220007562ef5).

That's what my instinct was. So far, I've gathered that beacon is using `window` under the hood and that's why the error. I've tried to circumvent that at several places by checking the if window is available. Still, no luck.

**Comment by bhaskarSingh** [comment]
Try asking in beacon dev group for a solution.

**Comment by manangouhari** [comment]
Traced this back to https://github.com/airgap-it/beacon-sdk/issues/108

File: src/components/NavBar/index.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
+import React, { useContext } from 'react';
+import { BeaconContext } from '../../context/beacon-context';
+import { Link, navigate } from 'gatsby';
+import Theme from 'src/assets/theme.svg';
+import userAtom from '../../atoms/user-atom';
+import isUserAtom from '../../atoms/is-user-atom';
+import { useAtom } from 'jotai';
+import { createUser } from '../../api';
+import { setUserId } from 'amplitude-js';
```
Comment (bhaskarSingh) [inline src/components/NavBar/index.js:9]: @manangouhari Please try removing the `amplitude-js` import.

Screenshot 2021-01-30 at 12 51 59 PM

Since amplitude is also using window object under the hood, that's why it's resulting in a build error and it's not being used here anyways.

Comment (bhaskarSingh) [reply src/components/NavBar/index.js:9]: Tried removing it & run a local build. It's working, yes!!!!

**Comment by bhaskarSingh** [comment]
Has the edge case where the user hasn't created an account taken care of? (ref: https://www.figma.com/proto/eUQ3EIhaGW2N8CkTu6Ndnm/BUIDL-Labs---Tezos---Cryptoverse-Wars---Final-Designs?node-id=910%3A6540&viewport=241%2C488%2C0.015625&scaling=min-zoom)

**Comment by manangouhari** [comment]
> Has the edge case where the user hasn't created an account taken care of? (ref: https://www.figma.com/proto/eUQ3EIhaGW2N8CkTu6Ndnm/BUIDL-Labs---Tezos---Cryptoverse-Wars---Final-Designs?node-id=910%3A6540&viewport=241%2C488%2C0.015625&scaling=min-zoom)

That case is same as when Thanos isn't available.

File: src/components/NavBar/index.js (hunk 1 above)
Comment (manangouhari) [reply src/components/NavBar/index.js]: Damn. How do I keep missing these small things. Need to up my game.

**Comment by manangouhari** [comment]
@bhaskarSingh please test it now.

**Comment by bhaskarSingh** [comment]
Suppose I have logged in, after which I uninstall the Thanos wallet. It still logs in even when the wallet is not present.

Screenshot 2021-01-31 at 8 41 24 PM
In case the wrong email id is entered. The above is shown, I think we need to update the copy here.

@harsh242 Can you also review it.

**Comment by bhaskarSingh** [comment]
> Has the edge case where the user hasn't created an account taken care of? (ref: https://www.figma.com/proto/eUQ3EIhaGW2N8CkTu6Ndnm/BUIDL-Labs---Tezos---Cryptoverse-Wars---Final-Designs?node-id=910%3A6540&viewport=241%2C488%2C0.015625&scaling=min-zoom)
That case is same as when Thanos isn't available.

Screenshot 2021-01-31 at 10 44 53 PM

It shouldn't be, Above I have installed Thanos but still it's telling me to install Thanos but rather it should prompt me to create a new account.

**Comment by harsh242** [comment]
> Suppose I have logged in, after which I uninstall the Thanos wallet. It still logs in even when the wallet is not present.
Ya ya, the same happened with me as well, I had disabled the extension.

> In case the wrong email id is entered. The above is shown, I think we need to update the copy here.
When does this come up?
When used an older email👇
image


> @harsh242 Can you also review the overall auth flow. It can be accessed at ` /auth`.
Same issues as you mentioned above.
Also to enhance the experience we can add loaders

**Comment by manangouhari** [comment]
> Screenshot 2021-01-31 at 8 41 24 PM
> In case the wrong email id is entered. The above is shown, I think we need to update the copy here.

Copy can't be changed. It's Magic's copy.

**Comment by manangouhari** [comment]
> > Has the edge case where the user hasn't created an account taken care of? (ref: https://www.figma.com/proto/eUQ3EIhaGW2N8CkTu6Ndnm/BUIDL-Labs---Tezos---Cryptoverse-Wars---Final-Designs?node-id=910%3A6540&viewport=241%2C488%2C0.015625&scaling=min-zoom)
> > That case is same as when Thanos isn't available.
> 
> Screenshot 2021-01-31 at 10 44 53 PM
> 
> It shouldn't be, Above I have installed Thanos but still it's telling me to install Thanos but rather it should prompt me to create a new account.

Need to update the copy to reflect that either install thanos or create an account. 
In the code I'm using a function provided by Thanos to check availability, it returns false on both conditions - not installed or account not created.

**Comment by manangouhari** [comment]
> Suppose I have logged in, after which I uninstall the Thanos wallet. It still logs in even when the wallet is not present.

Will debug this.

**Comment by bhaskarSingh** [comment]
TODO left #152
